src/SimpleSAML/Metadata/MetaDataStorageHandler.php

```diff
@@ -264,32 +264,46 @@ public function getPreferredEntityIdFromCIDRhint(string $set, string $ip): ?stri
      */
     public function getMetaDataForEntities(array $entityIds, string $set): array
     {
+        $startTime = microtime(true);
         $result = [];
+        // We are flipping the entityIds array in order to avoid constant iteration over it.
+        // Even if it becomes smaller over time.
+        // Still, after flipping all actions will be O(1)
+        $entityIdsFlipped = array_flip($entityIds);
         $timeUtils = new Utils\Time();
+
         foreach ($this->sources as $source) {
-            $srcList = $source->getMetaDataForEntities($entityIds, $set);
+            // Break iteration if the entityID List is empty
+            if (empty($entityIdsFlipped)) {
```

Comment:
This if-statement could be moved outside the foreach-loop.. If the variable is not empty, run the foreach..

Reply:
@tvdijen we need to break looping the sources as soon as the `entityIdsFlipped` variable get empty, check line 293, `unset($entityIdsFlipped[$key]);

Reply:
I agree with @ioigoume that the goal was to stop iterating through the sources once all the entityIds have been found.